File: Top K Frequent Words.cpp
class Solution {
public:
    vector<string> topKFrequent(vector<string>& words, int k) {
        unordered_map<string, int> m;
        for(string& word : words) 
            m[word]++;
        
        // comp(a, b) == true means a has LOWER priority than b
        auto comp = [&](string& a, string& b){
            if(m[a] == m[b])
                return a < b;
            return m[a] > m[b];
        };

        priority_queue<string, vector<string>, decltype(comp)> pq(comp);
        for(auto& [word, _] : m) {
            pq.push(word);
            if(pq.size() > k) 
                pq.pop();
        }
        vector<string> res(k);
        for(int i = k-1; i >= 0; i--) {
            res[i] = pq.top();
            pq.pop();
        }
        return res;
    }
};

File: Top K Frequent Words.py
class Solution:
    def topKFrequent(self, words: List[str], k: int) -> List[str]:
        d = defaultdict(int)
        for word in words:
            d[word] += 1
        
        heap = []
        for word, count in d.items():
            heapq.heappush(heap, (-count, word))
            
        return [heapq.heappop(heap)[1] for _ in range(k)]
